### PR TITLE
Create subspace_learning_lib.py

### PR DESCRIPTION
**Pull request by karalleyna**
## Steps

- [x] Implement functions for SGD and NUTS so that they have similar structure like build_sghmc_sampler etc. from sgmcmcjax
- [x] Implement subspace_learning so that it uses any algorithm to fit the model.
- [x] Add progressing bars to SGD and nuts
- [x] Instead of allowing only full batch gradient descent, allow SGD

## Figures    

image

image

### CODE REVIEW

File: scripts/subspace_learning_lib.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+
+def generate_random_basis(key, d, D):
+    projection_matrix = normal(key, shape=(d, D))
+    projection_matrix = projection_matrix / jnp.linalg.norm(projection_matrix, axis=-1, keepdims=True)
+    return projection_matrix
+
+
+@jit
+def adam_update(grads, params, mass, velocity, hyperparams):
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:61]: why not use optax?

File: scripts/subspace_learning_lib.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    predicted_class = jnp.argmax(logits, axis=-1)
+    return jnp.mean(predicted_class == targets), jnp.exp(logits)
+
+
+def data_stream(dataset, batch_size):
+    X, y = dataset
+    num_train = len(X)
+    num_batches, rem = divmod(num_train, batch_size)
+    num_batches += 1 if rem else 0
+    rng = np.random.RandomState(0)
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:99]: I believe it's considered bad practice to mix np.rng and jax.prng. I suggest passing in a jax.prng key.

File: scripts/subspace_learning_lib.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+    _, states = scan(one_step, (initial_state, rng_key), jnp.arange(num_samples))
+    return states
+
+
+def train_step(carry, i, dataloader, loss, hyperparams, callback):
+    batch = next(dataloader)
+    params, mass, velocity = carry
+    epoch_loss, grads = value_and_grad(loss)(params, batch)
+    params, mass, velocity = adam_update(grads, params, mass, velocity, hyperparams)
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:123]: this should be any optimizer, not just adam

File: scripts/subspace_learning_lib.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+def build_nuts_sampler(num_warmup, potential):
+    def nuts_sampler(rng_key, num_samples, initial_params):
+        initial_state = nuts.new_state(initial_params, potential)
+
+        kernel_generator = lambda step_size, inverse_mass_matrix: jit(nuts.kernel(
+            potential, step_size, inverse_mass_matrix))
+
+        stan_key, key = split(rng_key)
+
+        final_state, (step_size, inverse_mass_matrix), _ = stan_warmup.run(
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:137]: if num_warmup = 0, you should have defaiults for `inverse_mass_matrix = jnp.eye()` and `step_size = 1e-3` or whatever.

File: scripts/subspace_mcmc_demo.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    Relu,
+    Dense(num_actions),
+    LogSoftmax
+)
+
+key = jax.random.PRNGKey(42)
+init_key, key = split(key)
+_, params_full_init = init_random_params(init_key, input_shape=(-1, num_features))
+
+d = 10
```
Comment (murphyk) [inline scripts/subspace_mcmc_demo.py:100]: it would be good to make a plot of agreement vs d. Also the ground truth you are comparing to should be NUTS in the full space, without projection.

**Review by murphyk — CHANGES_REQUESTED** [review]

File: scripts/subspace_mcmc_demo.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+import jax.numpy as jnp
+
+from subspace_learning_lib import subspace_learning, loglikelihood, logprior, accuracy
+from subspace_learning_lib import build_sgd, build_nuts_sampler
+import pyprobml_utils as pml
+
+from sgmcmcjax.samplers import *
+
+def load_dataset(filename):
+    with open(filename, "rb") as f:
```
Comment (murphyk) [inline scripts/subspace_mcmc_demo.py:27]: rather than using pickle files, I suggest just using
```
from sklearn.datasets import fetch_openml
X, y = fetch_openml('shuttle', version=1, return_X_y=True)
```
Also, removing any reference  to `num_actions`, and pass in the RNG key.

File: scripts/subspace_learning_lib.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    def projected_logprior(params_subspace):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        return logprior_fn(params_full)
+
+    def projected_potential(params_subspace, batch):
+        X, y = batch
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        batch_loglikelihood = vmap(loglikelihood_fn, in_axes=(None, 0, 0, None))(params_pytree, X, y, predict_fn)
+        return -jnp.sum(batch_loglikelihood) - projected_logprior(params_subspace)
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:198]: The potential should be defined as
```
U(theta) = -log_prior(theta) -N/B * sum_{i=1}^B loglik(i) 
```
where N is the total size of the dataset, and B is the minibatch size.

File: scripts/subspace_learning_lib.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+    def projected_logprior(params_subspace):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        return logprior_fn(params_full)
+
+    def projected_potential(params_subspace, batch):
+        X, y = batch
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        batch_loglikelihood = vmap(loglikelihood_fn, in_axes=(None, 0, 0, None))(params_pytree, X, y, predict_fn)
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:197]: why not use `vmap projected_loglikelihood`?

File: scripts/subspace_learning_lib.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    else:
+        sampler = sampler_init_fn(**hyperparams, loglikelihood=projected_loglikelihood, \
+                                  logprior=projected_logprior, data=(dataset["X"][:, None, ...], dataset["y"]))
+        num_samples += num_warmup
+
+    initial_subspace_params = hyperparams["centering_value"] if "centering_value" in hyperparams else jnp.zeros((1, d))
+    sample_key, key = split(key)
+    sampler_output = sampler(sample_key, num_samples, initial_subspace_params)
+
+    final_subspace_params = tree_map(lambda x: x[num_warmup:].mean(axis=0), sampler_output)
```
Comment (murphyk) [inline scripts/subspace_learning_lib.py:239]: the sampler should return all the samples (in the full space), not just their mean

File: scripts/subspace_mcmc_demo.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+num_samples = 3000
+batch_size = 1024
+optimizer_params = {"lr": 1e-1,
+                    "beta_1": 0.9,
+                    "beta_2": 0.999,
+                    "epsilon": 1e-7}
+
+models = {"SGD": {"sampler_fn": build_sgd,
+                  "hyperparams": {"batch_size": batch_size,
+                                  "optimizer_params": optimizer_params
```
Comment (murphyk) [inline scripts/subspace_mcmc_demo.py:111]: for all your samplers, the `optimizer_params` field is outside of `hyperparams` field, why is SGD different?

**Review by murphyk — CHANGES_REQUESTED** [review]

File: scripts/subspace_mcmc_demo.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            ticks.append(f"NUTS - {k}")
+            i += 1
+
+    ax.set_yticks(range(i + 1))
+    ax.set_yticklabels(ticks)
+    ax.set_xlabel(xlabel, fontsize=12)
+    pml.save_fig(f'{filename}.pdf')
+    plt.show()
+
+pickle_name = './statlog-bandit-data.pkl'
```
[the remark this answers is not in the document]
Comment (karalleyna) [reply scripts/subspace_mcmc_demo.py:78]: [Done](https://github.com/probml/probml-data/commit/0725771b37c24b3c0ec882802e75e64ba9cb4fdd) :)

**Review by murphyk — CHANGES_REQUESTED** [review]
